lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -152,25 +152,37 @@ void HSICEstimatorImplementation::computePValuesPermutation() const
 {
   const SquareMatrix Wobs(computeWeightMatrix(outputSample_));
   PValuesPermutation_ = Point(inputDimension_);
+  Sample ShuffledSample;
+  Collection<Sample> shuffleCollection(permutationSize_);
+  Collection<SquareMatrix> weightMatrixCollection(permutationSize_);
+
+  for( UnsignedInteger b = 0; b < permutationSize_; ++b)
+  {
+    ShuffledSample = shuffledCopy(outputSample_);
+    shuffleCollection[b] = ShuffledSample;
+    weightMatrixCollection[b] = computeWeightMatrix(ShuffledSample);
+  }
 
   for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
   {
+
     const Sample xdim(inputSample_.getMarginal(dim));
     const Scalar HSIC_obs = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], Wobs);
-
+    Scalar HSIC_loc;
     UnsignedInteger count = 0;
+
     for( UnsignedInteger b = 0; b < permutationSize_; ++b)
     {
-      const Sample Yp(shuffledCopy(outputSample_));
-      const SquareMatrix W(computeWeightMatrix(Yp));
-      const Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
```

Comment:
you can keep variable initialization inside loop, this is no c89 :)

Reply:
Is this recommended? Is the goal to have the variables exist only within the scope of the for loop?